### DISCUSSION TITLE
Suggestions please for an enhanced audio player

### DISCUSSION THREAD

**Discussion by obdevel** [pschatzmann/arduino-audio-tools#257] [Q&A] [answered]
I'd like to create an enhanced audio player and I'd welcome any suggestions.

It would be a development of the SD->MP3->Mixer->I2S pipeline but with some specific requirements. I already have a basic player working on ESP32 + MAX98357 with 4 input streams.

- 'single-shot' samples would be added to the mix according to external events (e.g. a switch press) and just run to completion
- I want to loop one or more of the inputs, which I can do with a new Stream class. I'll just loop back to the beginning of the file when I reach EOF
- I'd like to cache MP3 files from SD into PSRAM (for performance), which I can do at startup and then access them as MemoryStreams

- can I change an input stream source at runtime (to a different file) by manipulating the mixer's streams vector ?
- is it possible to 'speed up' an input stream by 'dropping' some samples ? I think this would be a simple in->out copier class. Where in the pipeline could I do this ? After decoding but before the mixer ? Do you think this would have the desired effect ? (I expect there is a limit to this, before the audio quality is affected by the missing data).

Any thoughts or advice would be very welcome.

Thank you for such a useful library. So many possibilities !

**Comment by pschatzmann** [comment, marked as answer]
A couple of comments:
- The MemoryStream provides some looping (auto revind) functionality: just use setLoop(true). I think it would make sense to have a more generic Stream class which would work for other cases as well. 
- The SD is usually fast enough and does not need any caching. Please note that the MemoryStream supports PSRAM so you could try to copy the data.
-  'speed up' : I am not sure what you want to achieve. I suggest that you look into [Resampling](https://github.com/pschatzmann/arduino-audio-tools/wiki/Resampling) - If the intention is to get a Mikey Mouse effect you could just change the sample rate.
- Mixing: There is some basic functionality but I was struggling with with the Mixing on the output side lately

**Comment by obdevel** [comment]
Thank you Phil.

By 'speed up' I was thinking about e.g. a steam engine, and changing the sample 'speed' to match the engine's speed. I think resampling would be like speeding up a tape, which would result in a pitch change (the 'mickey mouse' effect, or nitrous oxide gas!). They would likely be limits to dropping samples, after which the sound quality would be affected, but it would reduce the number of source files required.

But I can experiment with this. I'll also experiment with manipulating the mixer's streams and weights vectors at runtime.

Thanks again.

**Comment by obdevel** [comment]
I've added a couple of methods to the InputMixer class (addAt() and remove()) which manipulate the streams and weights vectors. This works fine and I can add and remove streams at runtime. There is a brief 'splat' artefact when removing the last stream or calling the end() method, but my project would always have at least one looped stream running and I can reduce the volume to 0.0 if required.

My test pipeline is: 4x SineWaveGenerator -> mixer -> volume control -> I2S

**Comment by pschatzmann** [comment]
You could incrementally reduce the weight until 0 and remove only if it is at 0. This way you could avoid any artefacts

**Comment by obdevel** [comment]
Thanks, I'll try that.

Another question, if I may: which class do I need to connect one or more MP3 streams to a mixer ? My intended pipeline is:

MemoryStream -> MP3DecoderHelix -> <???> -> InputMixer -> VolumeStream -> I2SStream

Which class is <???> ? I've been testing so far with SineWaveGenerator -> GeneratedSoundStream -> InputMixer ...

**Comment by pschatzmann** [comment]
You would use an OutputMixer instead of an InputMixer, but I think there is not enough memory to decode more then one mp3
